### PR TITLE
Cut off `nose`

### LINKED ISSUE #1140: Python 3.12 breaks `nose`; all `nose` based testing should be removed
To be honest, we should have removed the `nose` requirement when we switched to pytest. But there were enough legacy tests that it was a headache to switch. Now it becomes actually important.

Apparently the main issue is that Python 3.12 [removed the `imp` module](https://docs.python.org/3.12/whatsnew/3.12.html), so `nose` now raises import errors. The best solution, by far, is for us to finally get rid of nose-based testing.

### PR DESCRIPTION
**Pull request by dwhswenson**
Resolves #1140, and will allow us to enable Python 3.12 testing (at least in minimal testing; unsure whether some upstreams are caught up yet).

This is a decent bit of tedious work (which is why it hasn't been done yet.) I can facilitate it with some vim macros, but there will be edge cases, so results from macros need to be manually validated. My guess is that I'll do it in the following steps:

- [x] Replace `assert_equal(a, b)` with `assert a == b`
- [x] Replace `assert_not_equal(a, b)` with `assert a != b`
- [x] Replace `raise SkipTest()` with `pytest.skip()`
- [x] Replace `@raises(...)` with `with pytest.raises(...)`
- [x] Replace `assert_almost_equal`, probably with things from `numpy.testing`
- [x] Replace various methods in `test_helpers` that are built on `nose`
- [x] Clean up any remaining bits that aren't included above
- [x] Remove `nose` from the build process

Adding Python 3.12 to the CI matrix will come in a later PR.

### CODE REVIEW

**Comment by dwhswenson** [comment]
This is ready for review and comment. Without review, I will merge this after at least 24 hours, no sooner than Fri 26 Apr 02:00 GMT (Thu 25 Apr 21:00 my local).

This is a huge PR in terms of changes, but all of them are individually small and repetitive. I converted legacy `nose` to `pytest` with the help of a number of vim macros. But do at least take a look at the many nose-based puns I've been making in commit messages, as well as in the PR title/branch name. Otherwise, like most of my jokes, these will be spoken into the void, only for my own amusement.

**Comment by sroet** [comment]
> But do at least take a look at the many nose-based puns I've been making in commit messages, as well as in the PR title/branch name. Otherwise, like most of my jokes, these will be spoken into the void, only for my own amusement.

Thanks for the heads up, I do appreciate them :grinning: Will walk through the PR and see how far I get

File: openpathsampling/tests/test_pathsimulator.py
Comment (sroet) [inline openpathsampling/tests/test_pathsimulator.py:722]: prefer no backslash
```suggestion
        assert (sim.flux_events[(state, alpha)]
                == expected_flux_events[(state, alpha)])
        assert (sim.flux_events[(state, beta)]
                == expected_flux_events[(state, beta)])
```

File: openpathsampling/tests/test_pathsimulator.py
Comment (sroet) [inline openpathsampling/tests/test_pathsimulator.py:744]: prefer no backslash
```suggestion
        assert (sim.flux_events[(state, interface)]
                == expected_flux_events[(state, interface)])
```

File: openpathsampling/tests/test_range_logic.py
Comment (sroet) [inline openpathsampling/tests/test_range_logic.py:58]: prefer no backslash
```suggestion
        assert (periodic_range_and(0.1, 0.4, 0.3, 0.2)
                == [(0.1, 0.2), (0.3, 0.4)])
```

File: openpathsampling/tests/test_storage.py
Comment (sroet) [inline openpathsampling/tests/test_storage.py:81]: prefer no backslash
```suggestion
        assert (self.simplifier.to_json(self.template_snapshot.topology)
                == self.simplifier.to_json(loaded_topology))
```

File: openpathsampling/tests/test_storage.py
Comment (sroet) [inline openpathsampling/tests/test_storage.py:68]: what is the reason that this test is skipped?

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:187]: prefer no backslash
```suggestion
            assert (weighted_trajs[ensembles_AB[ens]][self.trajs_AB[traj]]
                    == result)
            assert (weighted_trajs[ensembles_BA[ens]][self.trajs_BA[traj]]
                    == result)
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:196]: prefer no backslash (can't do suggestions due to unaltered stuff in between)

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:338]: prefer no backslash
```suggestion
        assert (set(minus_ensemble_to_mover.keys())
                == set(network.minus_ensembles))
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:433]: prefer no backslash
```suggestion
        assert (default_histogrammer.hist_parameters
                == {'bin_width': 5, 'bin_range': (0, 1000)})
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:683]: prefer no backslash
```suggestion
        assert (set(pair for pair in self.mistis_transition_dict)
                == set(pair for pair in self.mstis_transition_dict))
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:691]: prefer no backslash
```suggestion
        assert self.mstis_transition_dict[(self.state_A, self.state_B)] == 1
        assert (self.mstis_transition_dict[(self.state_A, self.state_B)]
                == self.mistis_transition_dict[(self.state_A, self.state_B)])
        assert self.mstis_transition_dict[(self.state_B, self.state_A)] == 2
        assert (self.mstis_transition_dict[(self.state_B, self.state_A)]
                == self.mistis_transition_dict[(self.state_B, self.state_A)])
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:702]: prefer no backslash
```suggestion
        assert (self.mistis_transition_dict[mistis_AB]
                == self.mstis_transition_dict[mstis_AB])
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:828]: prefer no backslash
```suggestion
            assert (self.mistis_analysis.state_fluxes(state)
                    == {(state, innermost): 0.1})
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:834]: prefer no backslash
```suggestion
            assert (self.mstis_analysis.state_fluxes(state)
                    == {(state, innermost): 0.1})
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:938]: prefer no backslash
```suggestion
                assert (expected_mistis.loc[(iface, state)]
                        == mistis_ctp.loc[(iface, state)])
```

File: openpathsampling/tests/test_tis_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_tis_analysis.py:949]: prefer no backslash
```suggestion
                assert (expected_mstis.loc[(iface, state)]
                        == mstis_ctp.loc[(iface, state)])
```

File: openpathsampling/tests/test_tools.py
Comment (sroet) [inline openpathsampling/tests/test_tools.py:19]: prefer no backslash
```suggestion
    assert (pretty_print_seconds(179990)
            == "2 days 1 hour 59 minutes 50 seconds")
```

File: openpathsampling/tests/test_tools.py
Comment (sroet) [inline openpathsampling/tests/test_tools.py:26]: prefer no backslash
```suggestion
    assert (pretty_print_seconds(179990, n_labels=3)
            == "2 days 2 hours 0 minutes")
```

File: openpathsampling/tests/test_tools.py
Comment (sroet) [inline openpathsampling/tests/test_tools.py:36]: prefer no backslash
```suggestion
    assert (pretty_print_seconds(93784, separator=", ")
            == "1 day, 2 hours, 3 minutes, 4 seconds")
```

File: openpathsampling/tests/test_tools.py
Comment (sroet) [inline openpathsampling/tests/test_tools.py:40]: prefer no backslash
```suggestion
    assert (progress_string(0, 100, 10)
            == "Starting simulation...\nWorking on first step\n")
```

File: openpathsampling/tests/test_toy_dynamics.py
Comment (sroet) [inline openpathsampling/tests/test_toy_dynamics.py:366]: prefer no backslash
```suggestion
        assert (self.sim.velocities[0] - init_vel[0]
                != self.sim.velocities[1] - init_vel[1])
```

File: openpathsampling/tests/test_trajectory_transition_analysis.py
Comment (sroet) [inline openpathsampling/tests/test_trajectory_transition_analysis.py:263]: I prefer no backslash and with parenthesis instead (can't due suggestion due to he unaltered line)

File: openpathsampling/tests/test_transitions.py
Comment (sroet) [inline openpathsampling/tests/test_transitions.py:114]: prefer no backslash
```suggestion
        assert (minus_sides_summary(self.traj_axaxa, minus)
                == {"in" : [5], "out" : [4]})
        assert (minus_sides_summary(self.traj_axaxa, minus)
                != {"in" : [5], "out" : [3]})
        assert (minus_sides_summary(self.traj_aixiaixia, minus)
                == {"in" : [5], "out" : [6] })
        assert (minus_sides_summary(self.traj_aixixiaxia, minus)
                == {"in" : [5], "out" : [10] })
```

File: openpathsampling/tests/test_transitions.py
Comment (sroet) [inline openpathsampling/tests/test_transitions.py:123]: prefer no backslash
```suggestion
        assert (minus_sides_summary(self.traj_axaxa, minus)
                == {"in" : [5], "out" : [4]})
        assert (minus_sides_summary(self.traj_aixiaixia, minus)
                == {"in" : [6], "out" : [4] })
        assert (minus_sides_summary(self.traj_aixixiaxia, minus)
                == {"in" : [5], "out" : [8] })
```

File: openpathsampling/tests/test_transitions.py
Comment (sroet) [inline openpathsampling/tests/test_transitions.py:140]: prefer no backslash
```suggestion
        assert (minus_sides_summary(self.traj_axaxaxaxa, minus)
                == {"in" : [2, 3, 4], "out" : [1, 2, 3]})
        assert (minus_sides_summary(self.traj_aixixiaixaxiaixia, minus)
                == {"in" : [4, 1, 5], "out" : [6, 2, 3]})
        assert (minus_sides_summary(self.traj_aixixiaixaxiaixia, minus)
                != {"in" : [1, 4, 5], "out" : [6, 2, 3]})
```

**Review by sroet — CHANGES_REQUESTED** [review]
> This is a huge PR in terms of changes, but all of them are individually small and repetitive. 

This makes it pretty easy to review, and unfortunately also means my many comments are also pretty repetative

To summarize:
Couple question about a test being skipped after it has already completed
Couple style nitpicks
Couple comments about unnecessary backslashes 
Couple comments about me preferring no backslashes (feel free to ignore these)
One test case went missing (this is the main one that needs to be changed)

Other things I noticed:
- you sometimes go for `assert x` and sometimes for `assert x is True` from `assert_equal(x, True)`, either one is fine by me (while I think `assert x` is technically the better replacement)

Other things I checked:
- I did run a `grep -R "nose" ./` on this branch and it only matched a `.git` object

LGTM otherwise

File: openpathsampling/tests/test_ensemble.py
Comment (sroet) [reply openpathsampling/tests/test_ensemble.py:2721]: @dwhswenson This is the main comment that needs to be handled for approval

File: openpathsampling/tests/test_bias_function.py
Comment (dwhswenson) [reply openpathsampling/tests/test_bias_function.py]: yep, the dead giveaway is the bunch of blank lines before it. That's usually a sign that I had put a skip in place while I was developing, so that it wouldn't just pass, but after a few refactors of the test it fell below my screen.

File: openpathsampling/tests/test_ensemble.py
Comment (dwhswenson) [reply openpathsampling/tests/test_ensemble.py:1438]: > This skip should probably be split out into a separate test (can't select the line above it, which also should be removed)

agreed

File: openpathsampling/tests/test_ensemble.py
Comment (dwhswenson) [reply openpathsampling/tests/test_ensemble.py:1598]: > are these backslashes required withing parenthesis?

Not required, totally a case of my doing 1000 changes absentmindedly while listening to something/someone else.

File: openpathsampling/tests/test_helpers.py
Comment (dwhswenson) [reply openpathsampling/tests/test_helpers.py:284]: > Did you want to handle this TODO still in this PR?

nope; but I should be keeping a list of relatively low-effort tech debt remediation like this... the stuff that's easy to do without a lot of focus.

File: openpathsampling/tests/test_movescheme.py
Comment (dwhswenson) [reply openpathsampling/tests/test_movescheme.py:1206]: > I prefer parenthesis without backslash

To be honest, so do I.... but this was outside the realm of "pure macro" and backslashes were fewer keystrokes. Happy to take the changes!

**Review by dwhswenson — COMMENTED** [review]
First pass reply; and accepting most (if not all?) suggestions. Then I'll do a second pass to address the rest.

File: openpathsampling/tests/test_storage.py
Comment (dwhswenson) [reply openpathsampling/tests/test_storage.py:68]: looks like it is a test that @jhprinz started 8 years ago and never quite finished? It tries to load a template topology, but (1) never stores it; (2) netcdfplus storages don't seem to have the `template` attribute. Leaving as-is for now, it'll probably get yanked as part of a larger cleanup in the future.

**Review by dwhswenson — COMMENTED** [review]
Now for one more pass through to see what points remain unresolved....

File: openpathsampling/tests/test_helpers.py
Comment (dwhswenson) [reply openpathsampling/tests/test_helpers.py:284]: Added a [Tech Debt Remediation project](https://github.com/orgs/openpathsampling/projects/1/views/1). I find projects are better for small task that can be saved for a while; doesn't clutter issues, doesn't spam people; gives some categorization so it doesn't get as lost in the shuffle.

This is the first task on that project table.

**Comment by dwhswenson** [comment]
Changes should be all addressed! Take a look when you can, @sroet, and mark it green if ready!

Replies to a couple other points:

> you sometimes go for assert x and sometimes for assert x is True from assert_equal(x, True), either one is fine by me (while I think assert x is technically the better replacement)

Yeah, I agree. First macro I added was (of course) `assert_equal(X, Y)` to `assert X == Y`. Eventually I saw enough of the `assert_equal(X, True)` that I made a macro to make it into `assert X` (and `assert not X` for `False`). Which one you see depends on how far into the process I was when I hit that file.

> I did run a grep -R "nose" ./ on this branch and it only matched a .git object

`git ls-files | xargs grep "nose"`  😉 I have far too many random untracked files in my OPS directory.... (But that's how I found the random mention in an ipynb).

File: openpathsampling/tests/test_storage.py
Comment (sroet) [reply openpathsampling/tests/test_storage.py:68]: 👍 just wanted to make sure it was still a conscious decision

**Review by sroet — APPROVED** [review]
LGTM! thanks for handling the comments (and the useful note in the batch commit)

**Pull request merged**
